libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -789,67 +799,410 @@ bool GCS_MAVLINK::handle_mission_item(mavlink_message_t *msg, AP_Mission &missio
+        next_deferred_message_to_send = i;
+    }
+    if (repeating_message_found) {
+        chan_is_streaming |= (1U<<(chan-MAVLINK_COMM_0));
+    } else {
+        chan_is_streaming &= ~(1U<<(chan-MAVLINK_COMM_0));
+    }
+}
+
+uint32_t GCS_MAVLINK::reschedule_interval(deferred_message_t &deferred) const
+{
+    uint32_t interval_ms = deferred.interval_ms;
+
+    // never slow down heartbeats:
+    if (deferred.id == MSG_HEARTBEAT) {
+        return interval_ms;
+    }
+
+    // add in adjustments for streamrate-slowdown (e.g. based
+    // on feedback from telemetry radio on its state).
+    // slowdown is basically 50ths of a second
+    interval_ms += stream_slowdown * 20;
+
+    // slow most messages down if we're transfering parameters or
+    // waypoints:
+    if (_queued_parameter && deferred.id != MSG_NEXT_PARAM) {
+        // we are sending parameters, penalize everbody else:
+        interval_ms *= 4;
+    } else if (waypoint_receiving && deferred.id != MSG_NEXT_WAYPOINT) {
+        interval_ms *= 4;
     }
+
+    return interval_ms;
 }
 
 void GCS_MAVLINK::retry_deferred()
 {
-    push_deferred_messages();
+    if (next_deferred_message_to_send == no_deferred_message_to_send) {
+        point_next_deferred_message_to_next_message_to_send();
+    }
+
+    const uint32_t start = AP_HAL::millis();
+    gcs().set_out_of_time(false);
+    while (AP_HAL::millis() - start < 5) { // spend a max of 5ms sending messages.  This should never trigger - out_of_time() should become true
+        if (gcs().out_of_time()) {
+            break;
+        }
+        if (next_deferred_message_to_send == no_deferred_message_to_send) {
+            // no messages to send *at all*.
+            break;
+        }
+        deferred_message_t &deferred = deferred_messages[next_deferred_message_to_send];
+        if (deferred.send_not_before_ms > start) {
+            // next message to send isn't due to be sent yet
+            break;
+        }
+#if DEBUG_SEND_MESSAGE_TIMINGS
+        uint32_t start_send_message_us = AP_HAL::micros();
+#endif
+        if (!try_send_message(deferred.id)) {
+            // didn't fit in buffer...
+#if DEBUG_SEND_MESSAGE_TIMINGS
+            try_send_message_stats.no_space_for_message++;
+#endif
+            break;
+        }
+#if DEBUG_SEND_MESSAGE_TIMINGS
+        uint32_t delta_us = AP_HAL::micros() - start_send_message_us;
+        if (delta_us > try_send_message_stats.longest_time_us) {
+            try_send_message_stats.longest_time_us = delta_us;
+            try_send_message_stats.longest_id = deferred.id;
+        }
+#endif
+        if (deferred.interval_ms) {
+            // reschedule message
+            const uint32_t interval_ms = reschedule_interval(deferred);
+            uint32_t time_basis = deferred.send_not_before_ms;
+            if (time_basis + interval_ms < start) {
+                // very not good; we're getting behind...
+                time_basis = start;
```

Comment:
Well - this is a reset, so yeah, your average was too high :-)

What I'm trying to do here is to keep things regular (so e.g. a 500ms timing glitch doesn't result in all of your messages shifting along by 500ms).  The "if" here is a sanity check - it shouldn't be triggered in the best of worlds.

What we really should *also* be doing here is adjusting stream_slowdown and (somehow) warning the user that they're asking too much of the link.  I'm not sure where we would leak the value in stream_slowdown out, however, and we may end up fighting with the radio-feedback adjustments of the same variable...

Reply:
> Well - this is a reset, so yeah, your average was too high :-)

I don't quite agree with that. This might be a one time only where a message was so late that it should have sent it again - that doesn't mean that the average is too high. Also, even when the average is too high this isn't lowering it down, it's simply changing the base time for the interval.

> What I'm trying to do here is to keep things regular (so e.g. a 500ms timing glitch doesn't result in all of your messages shifting along by 500ms).

I would say this doesn't keep them regular as it is changing the time at which a message is sent. Example:

- message interval is 5ms
- it's supposed to sent at time 0, 5, 10, 15, 20, 25, 30, etc.
- it fails to send at 5 and only sends it at 12
- with this code everything will be changed and message will be sent at 17, 22, 27, 32; my proposal keeps message to be sent at 15, etc.

I also don't get your comment about all messages being shifted, can you give an example?